File: firmware/src/core/compression.cpp
#include "compression.h"

namespace compression {

size_t rle_encode(const uint8_t *in, size_t in_len, uint8_t *out, size_t out_cap)
{
    if (!in || !out) return 0;
    size_t oi = 0;
    size_t i = 0;
    while (i < in_len) {
        uint8_t val = in[i];
        size_t run = 1;
        while (i + run < in_len && in[i + run] == val && run < 255) run++;
        if (oi + 2 > out_cap) return 0; // not enough room
        out[oi++] = (uint8_t)run;
        out[oi++] = val;
        i += run;
    }
    return oi;
}

size_t rle_decode(const uint8_t *in, size_t in_len, uint8_t *out, size_t out_cap)
{
    if (!in || !out) return 0;
    if (in_len % 2 != 0) return 0; // must be count/value pairs
    size_t oi = 0;
    for (size_t i = 0; i + 1 < in_len; i += 2) {
        uint8_t count = in[i];
        uint8_t val = in[i + 1];
        if (oi + count > out_cap) return 0;
        for (uint8_t c = 0; c < count; c++) out[oi++] = val;
    }
    return oi;
}

size_t delta_encode(const uint8_t *in, size_t in_len, uint8_t *out, size_t out_cap)
{
    if (!in || !out) return 0;
    if (out_cap < in_len) return 0;
    if (in_len == 0) return 0;
    out[0] = in[0];
    for (size_t i = 1; i < in_len; i++) {
        out[i] = (uint8_t)(in[i] - in[i - 1]);
    }
    return in_len;
}

size_t delta_decode(const uint8_t *in, size_t in_len, uint8_t *out, size_t out_cap)
{
    if (!in || !out) return 0;
    if (out_cap < in_len) return 0;
    if (in_len == 0) return 0;
    out[0] = in[0];
    for (size_t i = 1; i < in_len; i++) {
        out[i] = (uint8_t)(out[i - 1] + in[i]);
    }
    return in_len;
}

} // namespace compression

File: pi/compression.py
"""Lightweight compression — must match firmware/src/core/compression.*

  RLE:   [count][value] pairs (count 1..255)
  Delta: first byte verbatim, then intra-buffer byte differences

Both are symmetric and bounds-safe. Compression auto-enables for payloads
of at least COMPRESS_THRESHOLD bytes.
"""

COMPRESS_THRESHOLD = 32


def should_compress(n: int) -> bool:
    return n >= COMPRESS_THRESHOLD


def rle_encode(data: bytes) -> bytes:
    out = bytearray()
    i = 0
    n = len(data)
    while i < n:
        val = data[i]
        run = 1
        while i + run < n and data[i + run] == val and run < 255:
            run += 1
        out.append(run)
        out.append(val)
        i += run
    return bytes(out)


def rle_decode(data: bytes) -> bytes:
    if len(data) % 2 != 0:
        raise ValueError("RLE stream must be count/value pairs")
    out = bytearray()
    for i in range(0, len(data), 2):
        out.extend(bytes([data[i + 1]]) * data[i])
    return bytes(out)


def delta_encode(data: bytes) -> bytes:
    if not data:
        return b""
    out = bytearray(len(data))
    out[0] = data[0]
    for i in range(1, len(data)):
        out[i] = (data[i] - data[i - 1]) & 0xFF
    return bytes(out)


def delta_decode(data: bytes) -> bytes:
    if not data:
        return b""
    out = bytearray(len(data))
    out[0] = data[0]
    for i in range(1, len(data)):
        out[i] = (out[i - 1] + data[i]) & 0xFF
    return bytes(out)
